File: src/java/pulse/cdm/patient/actions/SEForcedExhale.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package pulse.cdm.patient.actions;

import pulse.cdm.bind.PatientActions.ForcedExhaleData;

import pulse.cdm.properties.SEScalar0To1;
import pulse.cdm.properties.SEScalarTime;

public class SEForcedExhale extends SEConsciousRespirationCommand
{
  protected SEScalar0To1 expiratoryReserveVolumeFraction;
  protected SEScalarTime exhalePeriod;
  protected SEScalarTime holdPeriod;
  protected SEScalarTime releasePeriod;
  
  public SEForcedExhale()
  {
    expiratoryReserveVolumeFraction = null;
    exhalePeriod = null;
    holdPeriod = null;
    releasePeriod = null;
  }
  
  public void copy(SEForcedExhale other)
  {
    if(this==other)
      return;
    
    if(other.hasExpiratoryReserveVolumeFraction())
      getExpiratoryReserveVolumeFraction().set(other.expiratoryReserveVolumeFraction);
    if(other.hasExpiratoryReserveVolumeFraction())
      getExpiratoryReserveVolumeFraction().set(other.expiratoryReserveVolumeFraction);
    if(other.hasExpiratoryReserveVolumeFraction())
      getExpiratoryReserveVolumeFraction().set(other.expiratoryReserveVolumeFraction);
    if(other.hasExpiratoryReserveVolumeFraction())
      getExpiratoryReserveVolumeFraction().set(other.expiratoryReserveVolumeFraction);
  }
  
  public SEConsciousRespirationCommand getCopy()
  {
    SEConsciousRespirationCommand command = new SEForcedExhale();
    command.copy(this);
    
    return command;
  }
  
  public void reset()
  {
    super.reset();
    if (expiratoryReserveVolumeFraction != null)
      expiratoryReserveVolumeFraction.invalidate();
    if (exhalePeriod != null)
      exhalePeriod.invalidate();
    if (holdPeriod != null)
      holdPeriod.invalidate();
    if (releasePeriod != null)
      releasePeriod.invalidate();
  }
  
  public boolean isValid()
  {
    return hasExpiratoryReserveVolumeFraction() && hasExhalePeriod();
  }
  
  public static void load(ForcedExhaleData src, SEForcedExhale dst)
  {
    //SEConsciousRespirationCommand.load(src.getConsciousRespirationCommand(), dst);
    if(src.hasExpiratoryReserveVolumeFraction())
      SEScalar0To1.load(src.getExpiratoryReserveVolumeFraction(),dst.getExpiratoryReserveVolumeFraction());
    if(src.hasExhalePeriod())
      SEScalarTime.load(src.getExhalePeriod(),dst.getExhalePeriod());
    if(src.hasHoldPeriod())
      SEScalarTime.load(src.getHoldPeriod(),dst.getHoldPeriod());
    if(src.hasReleasePeriod())
      SEScalarTime.load(src.getReleasePeriod(),dst.getReleasePeriod());
  }
  
  public static ForcedExhaleData unload(SEForcedExhale src)
  {
    ForcedExhaleData.Builder dst = ForcedExhaleData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  
  protected static void unload(SEForcedExhale src, ForcedExhaleData.Builder dst)
  {
    //SEConsciousRespirationCommand.unload(src,dst.getConsciousRespirationCommand());    
    if (src.hasExpiratoryReserveVolumeFraction())
      dst.setExpiratoryReserveVolumeFraction(SEScalar0To1.unload(src.expiratoryReserveVolumeFraction));
    if (src.hasExhalePeriod())
      dst.setExhalePeriod(SEScalarTime.unload(src.exhalePeriod));
    if (src.hasHoldPeriod())
      dst.setHoldPeriod(SEScalarTime.unload(src.holdPeriod));
    if (src.hasReleasePeriod())
      dst.setReleasePeriod(SEScalarTime.unload(src.releasePeriod));
  }
  
  public boolean hasExpiratoryReserveVolumeFraction()
  {
    return expiratoryReserveVolumeFraction != null;
  }
  public SEScalar0To1 getExpiratoryReserveVolumeFraction()
  {
    if (expiratoryReserveVolumeFraction == null)
      expiratoryReserveVolumeFraction = new SEScalar0To1();
    return expiratoryReserveVolumeFraction;
  }
  
  public boolean hasExhalePeriod()
  {
    return exhalePeriod != null;
  }
  public SEScalarTime getExhalePeriod()
  {
    if (exhalePeriod == null)
      exhalePeriod = new SEScalarTime();
    return exhalePeriod;
  }
  
  public boolean hasHoldPeriod()
  {
    return holdPeriod != null;
  }
  public SEScalarTime getHoldPeriod()
  {
    if (holdPeriod == null)
      holdPeriod = new SEScalarTime();
    return holdPeriod;
  }
  
  public boolean hasReleasePeriod()
  {
    return releasePeriod != null;
  }
  public SEScalarTime getReleasePeriod()
  {
    if (releasePeriod == null)
      releasePeriod = new SEScalarTime();
    return releasePeriod;
  }
  
  public String toString()
  {
      return "Forced Exhale" 
          + "\n\tExpiratoryReserveVolumeFraction: " + (hasExpiratoryReserveVolumeFraction() ? getExpiratoryReserveVolumeFraction() : "Not Specified")
          + "\n\tExhalePeriod: " + (hasExhalePeriod() ? getExhalePeriod() : "Not Specified")
          + "\n\tHoldPeriod: " + (hasHoldPeriod() ? getHoldPeriod() : "Not Specified")
          + "\n\tReleasePeriod: " + (hasReleasePeriod() ? getReleasePeriod() : "Not Specified");
  }
}

File: src/cpp/cdm/patient/actions/SEForcedExhale.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "stdafx.h"
#include "patient/actions/SEForcedExhale.h"
#include "properties/SEScalar0To1.h"
#include "properties/SEScalarTime.h"
#include "io/protobuf/PBPatientActions.h"

SEForcedExhale::SEForcedExhale(Logger* logger) : SEConsciousRespirationCommand(logger)
{
  m_ExpiratoryReserveVolumeFraction = nullptr;
  m_ExhalePeriod = nullptr;
  m_HoldPeriod = nullptr;
  m_ReleasePeriod = nullptr;
}

SEForcedExhale::~SEForcedExhale()
{
  SAFE_DELETE(m_ExpiratoryReserveVolumeFraction);
  SAFE_DELETE(m_ExhalePeriod);
  SAFE_DELETE(m_HoldPeriod);
  SAFE_DELETE(m_ReleasePeriod);
}

void SEForcedExhale::Clear()
{
  SEConsciousRespirationCommand::Clear();
  INVALIDATE_PROPERTY(m_ExpiratoryReserveVolumeFraction);
  INVALIDATE_PROPERTY(m_ExhalePeriod);
  INVALIDATE_PROPERTY(m_HoldPeriod);
  INVALIDATE_PROPERTY(m_ReleasePeriod);
}

void SEForcedExhale::Copy(const SEForcedExhale& src)
{
  PBPatientAction::Copy(src, *this);
}

bool SEForcedExhale::IsValid() const
{
  return SEConsciousRespirationCommand::IsValid() && HasExpiratoryReserveVolumeFraction() && HasExhalePeriod();
}

bool SEForcedExhale::IsActive() const
{
  return SEConsciousRespirationCommand::IsActive();
}

bool SEForcedExhale::HasExpiratoryReserveVolumeFraction() const
{
  return m_ExpiratoryReserveVolumeFraction == nullptr ? false : m_ExpiratoryReserveVolumeFraction->IsValid();
}
SEScalar0To1& SEForcedExhale::GetExpiratoryReserveVolumeFraction()
{
  if (m_ExpiratoryReserveVolumeFraction == nullptr)
    m_ExpiratoryReserveVolumeFraction = new SEScalar0To1();
  return *m_ExpiratoryReserveVolumeFraction;
}
double SEForcedExhale::GetExpiratoryReserveVolumeFraction() const
{
  if (m_ExpiratoryReserveVolumeFraction == nullptr)
    return SEScalar::dNaN();
  return m_ExpiratoryReserveVolumeFraction->GetValue();
}

bool SEForcedExhale::HasExhalePeriod() const
{
  return m_ExhalePeriod == nullptr ? false : m_ExhalePeriod->IsValid();
}
SEScalarTime& SEForcedExhale::GetExhalePeriod()
{
  if (m_ExhalePeriod == nullptr)
    m_ExhalePeriod = new SEScalarTime();
  return *m_ExhalePeriod;
}
double SEForcedExhale::GetExhalePeriod(const TimeUnit& unit) const
{
  if (m_ExhalePeriod == nullptr)
    return SEScalar::dNaN();
  return m_ExhalePeriod->GetValue(unit);
}

bool SEForcedExhale::HasHoldPeriod() const
{
  return m_HoldPeriod == nullptr ? false : m_HoldPeriod->IsValid();
}
SEScalarTime& SEForcedExhale::GetHoldPeriod()
{
  if (m_HoldPeriod == nullptr)
    m_HoldPeriod = new SEScalarTime();
  return *m_HoldPeriod;
}
double SEForcedExhale::GetHoldPeriod(const TimeUnit& unit) const
{
  if (m_HoldPeriod == nullptr)
    return SEScalar::dNaN();
  return m_HoldPeriod->GetValue(unit);
}

bool SEForcedExhale::HasReleasePeriod() const
{
  return m_ReleasePeriod == nullptr ? false : m_ReleasePeriod->IsValid();
}
SEScalarTime& SEForcedExhale::GetReleasePeriod()
{
  if (m_ReleasePeriod == nullptr)
    m_ReleasePeriod = new SEScalarTime();
  return *m_ReleasePeriod;
}
double SEForcedExhale::GetReleasePeriod(const TimeUnit& unit) const
{
  if (m_ReleasePeriod == nullptr)
    return SEScalar::dNaN();
  return m_ReleasePeriod->GetValue(unit);
}

void SEForcedExhale::ToString(std::ostream &str) const
{
  str << "Forced Exhale";
  if (HasComment())
    str << "\n\tComment: " << m_Comment;
  str << "\n\tExpiratoryReserveVolumeFraction: "; HasExpiratoryReserveVolumeFraction() ? str << *m_ExpiratoryReserveVolumeFraction : str << "NaN";
  str << "\n\tExhalePeriod: "; HasExhalePeriod() ? str << *m_ExhalePeriod : str << "NaN";
  str << "\n\tHoldPeriod: "; HasHoldPeriod() ? str << *m_HoldPeriod : str << "NaN";
  str << "\n\tReleasePeriod: "; HasReleasePeriod() ? str << *m_ReleasePeriod : str << "NaN";
  str << std::flush;
}